File: week9/lab_java/bank/Customer.java
package bank;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Customer implements Serializable {
    private String name;
    private List<Account> accounts;

    public Customer(String name) {
        this.name = name;
        this.accounts = new ArrayList<>();
        this.accounts.add(new Account("Savings"));
        this.accounts.add(new Account("Loan"));
    }
    
    public Customer() {
        this.name = readName();
        this.accounts = new ArrayList<>();
        this.accounts.add(new Account("Savings"));
        this.accounts.add(new Account("Loan"));
    }

    private String readName() {        
        System.out.print("Create Accounts for Customer: ");
        return Config.IN.next();
    }

    public String getName(){
        return this.name;
    }

    public boolean match(String name){
        return this.name.equals(name);
    }
    
    private Account account(String type){
        for(Account a:accounts)
            if(a.hasType(type))
                return a;
        return null;
    }

    private void deposit() {
        Account savings = account("Savings");
        if(savings != null){
           savings.deposit(readAmount("deposit"));
        }else{
            System.out.println("No such account");
        }
    }

    private void withdraw() {
        double amount = readAmount("withdraw");
        Account savings = account("Savings");
        if(savings != null){
            if (savings.has(amount))
                savings.withdraw(amount);
            else
                System.out.println("Insufficient funds");
        }else{
            System.out.println("No such account");
        }
    }

    private void transfer() {
        double amount = readAmount("transfer");
        Account savings = account("Savings");
        Account loan = account("Loan");
        if(savings != null){
            if (savings.has(amount)){
                if(loan !=null){
                    savings.transfer(amount, loan);
                }else{
                    System.out.println("No such account");
                }
            }else{
                System.out.println("Insufficient funds");
            }
        }else{
            System.out.println("No such account");
        }
    }

    private double readAmount(String action) {
        System.out.print("Amount to "+action+": $");
        return Config.IN.nextDouble();
    }

    private void show() {        
        System.out.println(name+" bank statement: "+Config.DTF.format(Config.NOW));
        accounts.forEach(System.out::println);        
    }    
    
    @Override 
    public String toString(){
        String s = accounts.stream().map(Object::toString).collect(Collectors.joining(" | "));
        return name+"\t--> "+s; 
    }
    
    private char readChoice(){
        System.out.print("Customer menu (d/w/t/s/x): ");
        return Config.IN.next().charAt(0);
    }

    public void use() {
        System.out.println(name+" banking menu: "+Config.DTF.format(Config.NOW));
        char c;
        while((c = readChoice()) != 'x'){
            switch(c){
                case 'd': deposit(); break;
                case 'w': withdraw(); break;
                case 't': transfer(); break;
                case 's': show(); break;
                default: help(); break;
            }
        }
        System.out.println("Back to Bank menu");
    }
    
    private void help() {
        System.out.println("Menu options");
        System.out.println("d = deposit");
        System.out.println("w = withdraw");
        System.out.println("t = transfer");
        System.out.println("s = show");
        System.out.println("x = exit");
    }
}

File: week8/lab_py/bank/customer.py
from config import DTF,NOW
from account import Account

class Customer:

    def __init__(self):
        self.name = self.read_name()
        self.accounts = [Account("Savings"), Account("Loan")]

    def read_name(self):
        print("Enter Customer Name: ", end="")
        return input()

    def match(self, name):
        return self.name == name

    def account(self, account_type):
        for account in self.accounts:
            if account.has_type(account_type):
                return account
        return None

    def deposit(self):
        savings = self.account("Savings")
        if savings:
            savings.deposit(self.read_amount("deposit"))
        else:
            print("No such account")

    def withdraw(self):
        amount = self.read_amount("withdraw")
        savings = self.account("Savings")
        if savings:
            if savings.has_balance(amount):
                savings.withdraw(amount)
            else:
                print("Insufficient funds")
        else:
            print("No such account")

    def transfer(self):
        amount = self.read_amount("transfer")
        savings = self.account("Savings")
        loan = self.account("Loan")
        if savings:
            if savings.has_balance(amount):
                if loan:
                    savings.pay_to(amount, loan)
                else:
                    print("No such account")
            else:
                print("Insufficient funds")
        else:
            print("No such account")

    def read_amount(self, action):
        print(f'Amount to {action}: $', end="")
        return float(input())

    def show(self):
        print(f'{self.name} bank statement: {NOW.strftime(DTF)}')
        for account in self.accounts:
            print(account)

    def __str__(self):
        return f'{self.name}\t--> {' | '.join(map(str, self.accounts))}'

    def read_choice(self):
        print("Customer menu (d/w/t/s/x): ", end="")
        return input().strip().lower()

    def use(self):
        print(f'{self.name} banking menu: {NOW.strftime(DTF)}')
        choice = self.read_choice()
        
        while choice!='x':
            match choice:
                case 'd':
                    self.deposit()
                case 'w':
                    self.withdraw()
                case 't':
                    self.transfer()
                case 's':
                    self.show()
                case _:
                    self.help()

            choice = self.read_choice()

        print("Back to Bank menu")

    def help(self):
        print("Menu options")
        print("d = deposit")
        print("w = withdraw")
        print("t = transfer")
        print("s = show")
        print("x = exit")
